colibri/src/icons.rs

```diff
@@ -0,0 +1,192 @@
+fn extract_chain_id_and_address(identifier: &str) -> Option<(u64, String)> {
+    let parts: Vec<&str> = identifier.split('/').collect();
+    if parts.len() != 2 {
+        return None;
+    }
+
+    let chain_parts: Vec<&str> = parts[0].split(':').collect();
+    let address_parts: Vec<&str> = parts[1].split(':').collect();
+
+    if chain_parts.len() != 2 || address_parts.len() != 2 {
+        return None;
+    }
+
+    let chain_id = chain_parts[1].parse::<u64>().ok()?;
+    let address = address_parts[1].to_string();
+
+    Some((chain_id, address))
+}
+
+async fn query_token_icon_and_extension(
+    chain_id: u64,
+    address: &str,
+) -> Option<(Bytes, &'static str)> {
+    let client = Client::new();
+    let base_url = "https://raw.githubusercontent.com/SmolDapp/tokenAssets/refs/heads/main/tokens";
+    let lower_address = address.to_ascii_lowercase();
+
+    let urls = vec![
+        (
+            format!("{}/{}/{}/logo.svg", base_url, chain_id, lower_address),
+            "svg",
+        ),
+        (
+            format!("{}/{}/{}/logo-32.png", base_url, chain_id, lower_address),
+            "png",
+        ),
+    ];
+
+    for (url, extension) in urls {
+        debug!("Querying {}", url);
+        match client
+            .get(&url)
+            .timeout(Duration::from_secs(10))
+            .send()
+            .await
+        {
+            Ok(response) => {
+                let status = response.status();
+                if status.is_success() {
+                    if let Ok(bytes) = response.bytes().await {
+                        return Some((bytes, extension));
+                    }
+                } else if let Ok(text) = response.text().await {
+                    error!(
+                        "Got non success response status {} with text: {} ",
+                        status, text
+                    );
+                } else {
+                    error!("Got non success response status {}", status);
+                }
+            }
+            Err(e) => {
+                error!("Got error {} after querying {}", e, url);
+                continue;
+            }
+        }
+    }
+
+    None
+}
+
+fn url_encode_identifier(input: &str) -> String {
+    input
+        .chars()
+        .map(|c| match c {
+            'a'..='z' | 'A'..='Z' | '0'..='9' | '-' | '_' | '.' | '~' => c.to_string(),
+            _ => format!("%{:02X}", c as u8),
+        })
+        .collect()
+}
+
+// Find the icon in the path and if existing return its extension
+fn find_icon(path: &Path) -> Option<(Bytes, String)> {
+    if let Ok(entries) = std::fs::read_dir(path.parent().unwrap_or(Path::new("."))) {
+        for entry in entries.filter_map(Result::ok) {
+            if entry.path().file_stem() == path.file_stem() {
+                if let Some(extension) = entry.path().extension() {
+                    if let Some(ext_str) = extension.to_str() {
+                        if let Ok(contents) = std::fs::read(entry.path()) {
```

Comment:
io is not async. there is https://docs.rs/tokio/latest/tokio/fs/fn.read.html since you use this function in a async context

Reply:
So what you mean is that this uses blocking (sync) io in an async context, right? And with your suggestion I can also do async io?